services/repository/push.go

```diff
@@ -227,30 +227,14 @@ func pushUpdates(optsList []*repo_module.PushUpdateOptions) error {
 				}
 
 				oldCommitID := opts.OldCommitID
-				if oldCommitID == git.EmptySHA && len(commits.Commits) > 0 {
-					oldCommit, err := gitRepo.GetCommit(commits.Commits[len(commits.Commits)-1].Sha1)
-					if err != nil && !git.IsErrNotExist(err) {
-						log.Error("unable to GetCommit %s from %-v: %v", oldCommitID, repo, err)
-					}
-					if oldCommit != nil {
-						for i := 0; i < oldCommit.ParentCount(); i++ {
-							commitID, _ := oldCommit.ParentID(i)
-							if !commitID.IsZero() {
-								oldCommitID = commitID.String()
-								break
-							}
-						}
-					}
-				}
-
-				if oldCommitID == git.EmptySHA && repo.DefaultBranch != branch {
-					oldCommitID = repo.DefaultBranch
-				}
 
 				if oldCommitID != git.EmptySHA {
 					commits.CompareURL = repo.ComposeCompareURL(oldCommitID, opts.NewCommitID)
-				} else {
-					commits.CompareURL = ""
+				} else if len(commits.Commits) == 1 {
+					commits.CompareURL = fmt.Sprintf("%s/commit/%s", repo.FullName(), opts.NewCommitID)
+				} else if len(commits.Commits) > 1 {
+					oldCommitID = commits.Commits[len(commits.Commits)-1].Sha1
+					commits.CompareURL = fmt.Sprintf("%s/compare/%s^...%s", repo.FullName(), oldCommitID, opts.NewCommitID)
```

Comment:
Does it work well?


https://try.gitea.io/wxiaoguang/test/compare/30c92a1aca8062b3fe1f20c9a55704b18ab7078f^...1768d1abfe02d629fb37ad633cfecea174a4756f

And the char `^` is a special char which needs escaping.

![image](https://github.com/go-gitea/gitea/assets/2114189/30229a8f-6435-48ca-8a20-6da94d563d58)